googleapis/google-cloud-python
# fix(bigquery): use pyarrow fallback for improved schema detection

Closes #9206.

This PR adds additional logic to schema autodetection if `pyarrow` is available, and the types could not be detected for all columns.

### How to test
 - Make sure that pyarrow is installed
- Try loading data from a dataframe to a new table when the dataframe contains columns of type string, date, etc. (those that end up being the Pandas dtype "object"). Do not provide an explicit schema to the load job.

**Actual result (before the fix):**
The backend responds with an error (incompatible types).

**Expected result (after the fix):**
The load job completes successfully, and column types are correctly detected (strings, dates...). 

### Misc
The code might not work for _all_ scalar types, depending on how good the pyarrow's schema detection logic is. Which types to we care about the most?

We should update the test case to include all of them (we currently test strings and dates).
